lesson18/script.js

```diff
@@ -0,0 +1,186 @@
+        }
+        switchSlider(sliderData,currentIndex);
+    }, 3000);
+}
+
+const resetAutoPlay = (sliderData) => {
+    clearInterval(autoPlay);
+    autoSwitchSlider(sliderData);
+}
+
+const renderSliderItems = (sliderData) => {
+    const fragment = document.createDocumentFragment();
+    for(const image of sliderData){
+        const sliderItem = createAttributedElements({tag:"div",attrObj:{class:"slider__item"}});
+        const slideImage = createAttributedElements({tag:"img",attrObj:{class:"slide-image",src:image.img,alt:image.alt}});
+        
+        fragment.appendChild(sliderItem).appendChild(slideImage);
+    }
+
+    slider.appendChild(fragment);
+}
+
+const getPageNumber = (sliderData) => {
+    const pageNumber    = createAttributedElements({tag:"p",attrObj:{class:"page-number"},str:`/${sliderData.length}`});
+    const currentNumber = createAttributedElements({tag:"span",attrObj:{id:"js-currentNumber",class:"current-number"},str:"1"});
+
+    slider.appendChild(pageNumber).insertAdjacentElement("afterbegin",currentNumber);
+}
+
+const renderButtons = (sliderData) => {
+    const prevBtn = createAttributedElements({tag:"button",attrObj:{id:"js-prevBtn",class:"js-btn prev-btn slider-btn",disabled:true}});
+    const nextBtn = createAttributedElements({tag:"button",attrObj:{id:"js-nextBtn",class:"js-btn next-btn slider-btn"}});
+
+    slider.appendChild(prevBtn).insertAdjacentElement("afterend",nextBtn);
+    addClickEventForButton(sliderData);
+}
+
+const renderDotPagination = (sliderData) => {
+    const paginationList = createAttributedElements({tag:"ul",attrObj:{id:"js-dot-pagination",class:"dot-pagination"}});
+
+    const fragment = document.createDocumentFragment();
+    const length   = sliderData.length;
+    for(let i = 0; i < length; i++){
+        const paginationItem = createAttributedElements({tag:"li",attrObj:{class:"dot-pagination__item"}});
+        fragment.appendChild(paginationItem);
+    }
+
+    slider.appendChild(paginationList).appendChild(fragment);
+    addClickEventForDotPagination(sliderData);
+}
+
+let currentIndex = 0;
+const addClickEventForButton = (sliderData) => {
+    const sliderButtons = document.getElementsByClassName("js-btn");
+
+    for (const button of sliderButtons ){
+        button.addEventListener("click",function(){
+            this.id === "js-nextBtn" ? currentIndex++ : currentIndex--;
+
+            switchSlider(sliderData,currentIndex);
+            resetAutoPlay(sliderData);
+        });
+    }
+}
+
+const addClickEventForDotPagination = (sliderData) => {
+    const dotPaginetions   = document.getElementsByClassName("dot-pagination__item");
+    const arrayPaginations = [...dotPaginetions];
+    
+    for(const dotPagination of dotPaginetions ){
+        dotPagination.addEventListener("click",function(){
+            currentIndex = arrayPaginations.indexOf(this);
+
+            switchSlider(sliderData,currentIndex);
+            resetAutoPlay(sliderData);
+        });
+    }
+}
```

Comment:
IMO
I thought that the event should be given to the parent's` <ul>`, not every single button.
Because it would affect performance.

```suggestion
const addClickEventForDotPagination = (sliderData) => {
    const dotPaginations = document.getElementById("js-dot-pagination");
    const arrayPaginations = [...document.getElementsByClassName("dot-pagination__item")];
    
    dotPaginations.addEventListener("click", (e) => {
        currentIndex = arrayPaginations.indexOf(e.target);

        switchSlider(sliderData,currentIndex);
        resetAutoPlay(sliderData);
    });
}
```
※ Let me know if there is a better way to write within the event...

Reply:
I didn't know much about the delegation of the event, so thank you for letting me know!
I fixed.
https://github.com/nakagawamai/js-lesson/pull/20/commits/ee6fdc9841ca80ad125ec72784a67a41d0535b4f